### ISSUE TITLE
[AIR] Support functionality to stitch Preprocessor with Keras model

### ISSUE THREAD

**Issue by daikeshi** [ray-project/ray#24023]
### Description

The existing preprocessor used for feature transformation needs to be pickled then can be used for serving. If the model serving is using a different language, such as Java, it's almost impossible to use Ray's preprocessor for the feature transformation. It would be nice if the feature transformation graph can be added to the Keras model graph similar to what `TF.Transform` does. 

This is [an example](https://www.tensorflow.org/tfx/tutorials/transform/census#train_evaluate_the_model) of using `TF.Transform` on a Keras model. When building a Keras model, we can add an encoded input layer on top of model graph for parsing the raw features. 
```python
def build_keras_model(working_dir):
  inputs = build_keras_inputs(working_dir)

  encoded_inputs = encode_inputs(inputs)

  stacked_inputs = tf.concat(tf.nest.flatten(encoded_inputs), axis=1)
  output = tf.keras.layers.Dense(100, activation='relu')(stacked_inputs)
  output = tf.keras.layers.Dense(50, activation='relu')(output)
  output = tf.keras.layers.Dense(2)(output)
  model = tf.keras.Model(inputs=inputs, outputs=output)

  return model
```

`build_keras_inputs` and `encode_inputs` (see the example below) relies on the feature spec produced by `TF.transform`. 

```python
def build_keras_inputs(working_dir):
  tf_transform_output = tft.TFTransformOutput(working_dir)

  feature_spec = tf_transform_output.transformed_feature_spec().copy()
  feature_spec.pop(LABEL_KEY)

  # Build the `keras.Input` objects.
  inputs = {}
  for key, spec in feature_spec.items():
    if isinstance(spec, tf.io.VarLenFeature):
      inputs[key] = tf.keras.layers.Input(
          shape=[None], name=key, dtype=spec.dtype, sparse=True)
    elif isinstance(spec, tf.io.FixedLenFeature):
      inputs[key] = tf.keras.layers.Input(
          shape=spec.shape, name=key, dtype=spec.dtype)
    else:
      raise ValueError('Spec type is not supported: ', key, spec)

  return inputs

def encode_inputs(inputs):
  encoded_inputs = {}
  for key in inputs:
    feature = tf.expand_dims(inputs[key], -1)
    if key in CATEGORICAL_FEATURE_KEYS:
      num_buckets = tf_transform_output.num_buckets_for_transformed_feature(key)
      encoding_layer = (
          tf.keras.layers.CategoryEncoding(
              num_tokens=num_buckets, output_mode='binary', sparse=False))
      encoded_inputs[key] = encoding_layer(feature)
    else:
      encoded_inputs[key] = feature

  return encoded_inputs
```

Ray's preprocessor should consider providing the similar support, so users can add feature preprocessing layer to a Keras model.   

### Use case

With this feature, users can train a TF Keras model with Ray's Preprocessor in Python and serve it in JVM directly with untransformed raw features.

**Comment by richardliaw** [comment]
OK, looks like we need to do a bit of scoping in order to better understand what's needed here.

We should also build an example that demonstrates this.

**Comment by cszhu** [comment]
Hello! This P2 issue has seen no activity in at least 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
